### PR TITLE
Pool AddCompletions

### PR DESCRIPTION
**Pull request by ivankelly**
To avoid generating garbage in the fastpath.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/bookkeeper-precommit-pullrequest-docker/128/

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 1]
```diff
@@ -832,28 +836,7 @@ private void writeAndFlush(final Channel channel,
...
-                                              channel.remoteAddress());
-                                }
-                            } else {
-                                if (!(future.cause()
-                                      instanceof ClosedChannelException)) {
-                                    LOG.warn("Writing request {} to {} failed : ",
-                                             requestToString(request),
-                                             channel, future.cause());
-                                }
-                                errorOut(key);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: Are we dropping this errorOut in case of write error?

Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: @eolivelli Yes. Any write error will result in an exception on the channel, which will result in the channel being closed, which will result in all active keys being errored out.

Comment (eolivelli) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: @ivankelly do we have some sort of test case about this case ?
Anyway your explanation is very clear to me

**Comment by ivankelly** [comment]
Latest push is a straight rebase onto master to get rid of conflicts.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/bookkeeper-precommit-pullrequest-docker/141/

**Pull request closed without merging**